Gui: Add a context menu to select obstructed items (from RT's fork Pick Geometry)

tetektoza opened the pull request:
As the title says, this is from RT's fork. I only adjusted it a little bit and removed pie selection which was previously in the implementation.

To activate context menu - "G, G", then if it can't resolve edges vs faces, it adds QMenus on top to let user decide which one to select.

Note: this is not the full original feature, as I've mentioned it still needs to be partly ported, but since @FreeCAD/design-working-group still decides, this part is safe and we can adjust it later on, as it provides core mechanism. Originally this was called `Pick Geometry`. I also didn't try to thin out the code, so I left it for review.

Demo:

https://github.com/user-attachments/assets/aa8da407-4a26-40cf-b8b2-cb0f3fcfda94

Resolves: https://github.com/FreeCAD/FreeCAD/issues/21446

pure-machinery commented:
This seems ok, but without highlighting the hovered menu item and geometry at the same time this kind of loses the point. 

How would the menu behave when you have 200 geometric entities ?

tetektoza commented:
Since RT's fork diverged from us a little bit, I was fighting with this preselection highlight stuff for a moment. But anyways, enjoy.
@pure-machinery I've decided to add it now, since I was resolving CI problems anyways.
@FEA-eng Done, ready for testing.

Demo:

https://github.com/user-attachments/assets/2c94167a-8ed8-4e28-a564-644ac135999e

FEA-eng reviewed (approved):
Tested, works very well. Awesome feature!

kadet1090 commented:
I'll test it in terms of UX and check the code.

pierreporte commented:
I just built the branch and it doesn’t work for me. With the assembly example available on start page, the feature only lists assemblies (all named “Assembly”). The number of lines varies depending on where I click, but all of them selects the only assembly of the file. Also, there is no highlight of the geometry corresponding to the hovered line in the list.

It quite well with the Part Design example, and the new selection look you’re implementing will make it almost perfect. The only problems I have right now are:

* All the lines are bodies with the face name in parentheses. I think it should list possible selection as their actual geometry, so faces minimum, but better if it detects planar surfaces, cylindrical surfaces, etc.
* When triggering the feature while above an edge or vertex, it’s hard to have them listed. I think there is a problem of picking radius. For that kind of feature, I think that a large picking radius could be a good idea. If you do this, it could be a good idea to display it with a circle in the 3D view. In this case, moving the context menu out of that circle is necessary.
* Maybe a little bit different from first point, maybe make the body name in parentheses instead (basically inverting the display)?


In general, the shortcut doesn’t work. hitting <kbd>G</kbd> key twice does nothing. <kbd>V</kbd><kbd>F</kbd> works for fit all zoom so it’s not a problem with shortcuts in general.

pierreporte commented:
I didn’t expect the feature to work to add geometry to a selector, like fillet or chamfer in PD, but I think it could be useful to be able to do that.

@FreeCAD/design-working-group would this be a follow-up issue? If not, how should it work?

kadet1090 commented on src/Gui/Navigation/NavigationStyle.cpp:
```diff
@@ -1930,6 +1932,8 @@ void NavigationStyle::openPopupMenu(const SbVec2s& position)
         QAction *item = navMenuGroup->addAction(name);
         navMenu->addAction(item);
         item->setCheckable(true);
+        QByteArray data(style.first.getName());
+        item->setData(data);
```
```suggestion
        item->setData(QByteArray(style.first.getName()));
```

Or if that does not work it should be named with something less generic than `data`, for example `nameByteArray`

kadet1090 reviewed (changes requested):
Generally looks fine! I made few comments here and there, not all are crucial so no need to address absolutely everything. 

You also have some lint errors - there must be fixed before we merge. Nice to see that feature arriving in FreeCAD!

maxwxyz commented on src/Gui/Navigation/NavigationStyle.cpp:
```diff
@@ -1947,7 +1951,64 @@ void NavigationStyle::openPopupMenu(const SbVec2s& position)
+            pickAction = new QAction(cmd->getAction()->text(), contextMenu);
+            pickAction->setShortcut(cmd->getAction()->shortcut());
+        } else {
+            pickAction = new QAction(QObject::tr("Pick geometry"), contextMenu);
```
```suggestion
            pickAction = new QAction(QObject::tr("Select Under Cursor"), contextMenu);
```

maxwxyz commented on src/Gui/CommandView.cpp:
```diff
@@ -3959,6 +3962,114 @@ bool StdCmdAlignToSelection::isActive()
+  : Command("Std_PickGeometry")
+{
+    sGroup        = "View";
+    sMenuText     = QT_TR_NOOP("Pick geometry");
```
```suggestion
    sMenuText     = QT_TR_NOOP("Select Under Cursor");

```

@FreeCAD/design-working-group what do you think about that name for this tool? We currently do not use pick for selecting elements in FreeCAD. Other ideas: Cycle Selection; Select Geometry; or Context Select

tetektoza replied:
Done.

maxwxyz commented on src/Gui/CommandView.cpp:
```diff
@@ -3959,6 +3962,114 @@ bool StdCmdAlignToSelection::isActive()
+{
+    sGroup        = "View";
+    sMenuText     = QT_TR_NOOP("Pick geometry");
+    sToolTipText  = QT_TR_NOOP("Pick hidden geometries under the mouse cursor in 3D view.\n"
```
```suggestion
    sToolTipText  = QT_TR_NOOP("Displays a context menu at the mouse cursor to select overlapping or obstructed geometry in the 3D view.\n"

```

maxwxyz reviewed (changes requested):
Adjust UI strings according to guidelines

yorikvanhavre commented:
Waiting for @tetektoza to look at the suggestions/change requests

Reqrefusion commented:
@tetektoza could you take a look at this? It would be a great feature to have in 1.1.

Reqrefusion commented:
@tetektoza could you take a look at this? It would be a great feature to have in 1.1.

tetektoza commented:
Put this PR in Ready for Review as I've mostly done stuff I've wanted, just final touches and tomorrow it should have all of the commits so we will probably be able to ship it to 1.1. 

Some demos how it works now based on users input so I won't forget to upload them:

https://github.com/user-attachments/assets/9f843d26-9f29-4a6d-b071-429e4bad2024

https://github.com/user-attachments/assets/b2c6aec5-9a17-43a3-bfed-e9ac07bfc9d1

Note that I haven't applied all of the comments in this PR yet, so it will change a bit until I push out, but in terms of usability it shouldn't change a lot.

Also, you see so many items in the list because I've increased selection range to 200px just for the test.

tetektoza commented on src/Gui/Navigation/NavigationStyle.cpp:
```diff
@@ -1947,7 +1951,64 @@ void NavigationStyle::openPopupMenu(const SbVec2s& position)
+    if (separator && posAction)
+        contextMenu->insertSeparator(posAction);
+
+    QAction* used = contextMenu->exec(QCursor::pos());
+    if (used && navMenuGroup->actions().indexOf(used) >= 0 && used->isChecked()) {
+        QByteArray type = used->data().toByteArray();
+        QWidget* widget = viewer->getWidget();
+        while (widget && !widget->inherits("Gui::View3DInventor"))
+            widget = widget->parentWidget();
+        if (widget) {
+            // this is the widget where the viewer is embedded
+            Base::Type style = Base::Type::fromName((const char*)type);
+            if (style != this->getTypeId()) {
+                QEvent* event = new NavigationStyleEvent(style);
+                QApplication::postEvent(widget, event);
+            }
+        }
+        return;
+    }
```
Broke this up into several functions, I guess they are more descriptive.

tetektoza commented on src/Gui/Navigation/NavigationStyle.cpp:
```diff
@@ -1947,7 +1951,64 @@ void NavigationStyle::openPopupMenu(const SbVec2s& position)
+        // Execute the Pick Geometry command at this position
+        auto cmd = Application::Instance->commandManager().getCommandByName("Std_PickGeometry");
+        if (cmd && cmd->isActive()) {
+            cmd->invoke(0);
```
Added a comment with explanations but it's consistent throughout the app.

tetektoza commented on src/Gui/Selection/SelectionView.cpp:
```diff
@@ -704,4 +705,153 @@ void SelectionView::onEnablePickList()
+struct SubMenuInfo {
+    QMenu *menu = nullptr;
+    // Map from sub-object label to map from object path to element info
+    std::map<std::string, std::map<std::string, ElementInfo> > items;
```
Done.

tetektoza commented on src/Gui/Selection/SelectionView.cpp:
```diff
@@ -704,4 +705,153 @@ void SelectionView::onEnablePickList()
+            
+            QString text = QString::fromUtf8(sel.obj->Label.getValue());
+            if (!sel.element.empty()) {
+                text += QString::fromLatin1(" (") + QString::fromUtf8(sel.element.c_str()) + QString::fromLatin1(")");
```
Done.

tetektoza commented on src/Gui/CommandView.cpp:
```diff
@@ -3959,6 +3962,114 @@ bool StdCmdAlignToSelection::isActive()
+
+bool StdCmdPickGeometry::isActive()
+{
+    auto view = qobject_cast<View3DInventor*>(getMainWindow()->activeWindow());
+    return view != nullptr;
```
Done.

kadet1090 commented on src/Gui/Inventor/So3DAnnotation.cpp:
```diff
@@ -49,21 +55,72 @@ void SoDelayedAnnotationsElement::initClass()
+    }
+    
+    // Clear storage
+    elt->priorityPaths.clear();
     elt->paths.truncate(0);
```
Is there any difference between `truncate(0)` and `.clear()`? I did copy the truncate one from coin source, but `.clear()` seems like better way. If they are identical it woul'd be good to unify them.

kadet1090 commented on src/Gui/Inventor/So3DAnnotation.h:
```diff
@@ -64,6 +79,10 @@ class GuiExport SoDelayedAnnotationsElement: public SoElement
     }
 
+private:
+    // priority-aware paths
+    std::vector<PriorityPath> priorityPaths;
+    
     SoPathList paths;
```
Is there a way where we could have only `priorityPaths`?

kadet1090 commented on src/Gui/Selection/Selection.h:
```diff
@@ -409,6 +411,9 @@ class GuiExport SelectionSingleton : public Base::Subject<const SelectionChanges
      */
     void setVisible(VisibleState visible);
 
+    static bool isClarifySelectionActive();
+    static void setClarifySelectionActive(bool active);
```
same here

tetektoza replied:
Done.

kadet1090 commented on src/Gui/Selection/SelectionView.cpp:
```diff
@@ -704,4 +709,379 @@ void SelectionView::onEnablePickList()
+    menuArray.reserve(menus.size());
+    
+    for (const auto& category : preferredOrder) {
+        auto it = menus.find(category);
+        if (it != menus.end()) {
```
```suggestion
        if (auto it = menus.find(category); it != menus.end()) {
```

tetektoza replied:
Done.

kadet1090 commented on src/Gui/Selection/SelectionView.cpp:
```diff
@@ -704,4 +709,379 @@ void SelectionView::onEnablePickList()
+        }
+    }
+        
+    for (auto it : menuArray) {
+        auto &v = *it;
+        auto &info = v.second;
+        
+        if (info.items.empty()) {
+            continue;
+        }
+
+        info.menu = addMenu(QString::fromUtf8(v.first.c_str()));
+        bool groupMenu = shouldGroupMenu(info);
+
+        for (auto &vv : info.items) {
+            const std::string &label = vv.first;
+
+            for (auto &vvv : vv.second) {
+                auto &elementInfo = vvv.second;
+                
+                if (!groupMenu) {
+                    createFlatMenu(elementInfo, info.menu, label, v.first, selections);
+                } else {
+                    createGroupedMenu(elementInfo, info.menu, label, v.first, selections);
+                }
+            }
+        }
+    }
```
Can we have a bit better variable names here? `v`, `vv`, `vvv` are not that useful - frankly I don't know what they represent.

tetektoza replied:
Yes, added more descriptive ones.

kadet1090 commented on src/Gui/Selection/SelectionView.cpp:
```diff
@@ -704,4 +709,379 @@ void SelectionView::onEnablePickList()
+
+void SelectionMenu::addGeoFeatureTypes(App::DocumentObject* sobj, std::map<std::string, SubMenuInfo> &menus, std::set<std::string> &createdTypes)
+{
+    auto geoFeature = dynamic_cast<App::GeoFeature*>(sobj->getLinkedObject(true));
```
```suggestion
    auto geoFeature = freecad_cast<App::GeoFeature*>(sobj->getLinkedObject(true));
```

kadet1090 commented on src/Gui/Selection/SelectionView.cpp:
```diff
@@ -704,4 +709,379 @@ void SelectionView::onEnablePickList()
+
+bool SelectionMenu::shouldGroupMenu(const SubMenuInfo &info)
+{
+    if (info.items.size() > 20) {
```
That is magic number - can you give it some constexpr with proper name? Same thing for the `5` that is below.

tetektoza replied:
Added two `constexpr` for those thresholds.

kadet1090 commented on src/Gui/Selection/SelectionView.cpp:
```diff
@@ -704,4 +709,379 @@ void SelectionView::onEnablePickList()
+        if (!sel.element.empty()) {
+            text += QStringLiteral(" (%1)").arg(QString::fromUtf8(sel.element.c_str()));
+        } else if (elementType == "Object" && !sel.subName.empty() && sel.subName.back() == '.') {
+            text += QString::fromLatin1(" (Whole Object)");
```
Should be translated I think?

kadet1090 commented on src/Gui/Selection/SelectionView.cpp:
```diff
@@ -704,4 +709,379 @@ void SelectionView::onEnablePickList()
+        if (!sel.element.empty()) {
+            text = QString::fromUtf8(sel.element.c_str());
+        } else if (elementType == "Object" && !sel.subName.empty() && sel.subName.back() == '.') {
+            text = QString::fromLatin1("Whole Object");
```
Should be translated I think?

tetektoza replied:
Done.

kadet1090 commented on src/Gui/View3DInventorViewer.cpp:
```diff
@@ -229,9 +230,27 @@ while the progress bar is running.
-    ViewerEventFilter() = default;
+    ViewerEventFilter() : longPressTimer(new QTimer(this)) {
+        longPressTimer->setSingleShot(true);
+        longPressTimer->setInterval(1000); // after 1s of LMB press on viewport it gets toggled
```
Seems like something that should be configurable

kadet1090 commented on src/Mod/Part/Gui/SoBrepEdgeSet.cpp:
```diff
@@ -79,6 +87,20 @@ void SoBrepEdgeSet::GLRender(SoGLRenderAction *action)
         return;
+
+
+    if (Gui::Selection()
+            .isClarifySelectionActive() && !Gui::SoDelayedAnnotationsElement::isProcessingDelayedPaths 
+        && ((ctx && !ctx->hl.empty()) || viewProvider->isFaceHighlightActive())) {
```
Would it be possible to rewrite it in a more readable manner?

tetektoza replied:
Done, added separate variables.

kadet1090 reviewed (approved):
The code looks good, it also works fine. There are some linter warnings that should be fixed before merge.

maxwxyz reviewed (changes requested):
Great tool.

It does not always start or show elements (appears to work only randomly):

https://github.com/user-attachments/assets/d656085a-283d-44dd-8ed3-02ccc6772917




It does not work for some types it seems. It offers a blank next menu but still preselects in the 3D view:


https://github.com/user-attachments/assets/9b5fb53e-5eff-4eb9-b9d9-070e301f03c7

https://github.com/user-attachments/assets/f06fc794-19fd-4f15-af78-ab1fd7ecc790


Also, in RT it displays face/edge names. Here, all entries in the context menu have the object name, so the context menu has 10+ entries with the same name (sems to be for link objects):


https://github.com/user-attachments/assets/ab114e13-9450-4eaf-99bc-ad006138f05a

It seems to have issues with or does not work with
- Sketch Internal Faces (not showing up)
- Draft / BIM objects (arrays, dimensions,...)




https://github.com/user-attachments/assets/3b8a5f37-3a33-449f-aaf7-45b7e12eca14

Not sure if the pick radius is wrong, but if I start the tool on a preselected edge, that edge itself does not show up, but an attached edge to that:

https://github.com/user-attachments/assets/614c0c56-dbf3-441c-8b9f-9829a79339fb




The tool is currently in the end of the Tool menu.
I suggest putting it one spot before the unit converter.
Ideally give it a status tip / tooltip so it explains how to use the tool (Opens a context menu at the cursor position to select obstructed elements in the 3D view)
<img width="435" height="407" alt="grafik" src="https://github.com/user-attachments/assets/11add2a4-5aa8-4881-b9f7-5936bb8d077a" />

tetektoza commented:
So a couple of points, I will address them by video you've send:
Video 1 - this is due to right click behavior - we take the right mouse button position and check if something is under it, if yes, then Clarify Selection pops up in the context menu and then we check again after user selects. But in your case, after you click, nothing is in the position of your mouse after the click, so the menu doesn't pop up. I've fixed it by remembering right mouse button position in case user uses this menu instead of G, G or long-press LMB.

Video 2,3 - I removed flat menu creation for both of those categories - Other and Object. It was actually a bug, since i didn't test it on examples where there are so many objects that flat menus get created.

Video 4 - this is actually quite problematic, because arrays (on LinkBranch those are wires) and internal sketch faces seem to be objects that are somehow hidden upper in hierarchy. Let me add this in separate PR as I will also be adding CompoundSolid and Solids support that @FEA-eng requested, and they seem to be located in the same place.

Video 5 - this is the same as video 1. You right click, your mouse position is in one place, then in another after you move it to click "Clarify Selection" in the menu, and under current mouse position after that operation there is another object. So fix that fixes Video 1 should also fix it.

Last screenshot - added and the tooltip was already here.

++++++ I've added radius multiplier option, by default it is 5. It multiplies pixel radius that we have in preferences for selection - by default it is 5px. So using Clarify Selection brings it to 25px radius total, whereas default selection remains 5px.


Sample demo with current behavior and adressed points:

https://github.com/user-attachments/assets/cb7eaff7-58ba-43ce-8097-86e211f146eb

tetektoza commented:
@eblanshey hey, yeah. I plan to get back to it and add some support that was already requested in issues (I think your idea was requested recently as well). This one was just a bit exhausting to port and reimplement so I took a break and I'm mostly trying to fix bugs right now as it gives me a lot of context switching and short-time rewards for my brain :D But yeah, I will get back to it and you can expect the same functionality or similar to the existing one for other objects.
